packages/next/src/client/components/app-router.tsx

```diff
@@ -441,20 +488,83 @@ function Router({
         dispatch({
           type: ACTION_RESTORE,
           url: new URL(window.location.href),
-          tree: state.tree,
+          tree: state.__PRIVATE_NEXTJS_INTERNALS_TREE,
+        })
+      })
+    },
+    [dispatch]
+  )
+
+  // Ensure the canonical URL in the Next.js Router is updated when the URL is changed so that `usePathname` and `useSearchParams` hold the pushed values.
+  const applyUrlFromHistoryPushReplace = useCallback(
```

Comment:
(Non-blocking factoring nit) Since this is only ever called from the `useEffect` function below, you can define it inside the effect body. Can do the same for the onPopstate handler.

Reply:
Fixed!